### DISCUSSION TITLE
Security boundaries

### DISCUSSION THREAD

**Discussion by valoq** [roddhjav/apparmor.d#307] [Ideas]
As stated in the project objectives, the profiles in this repository are designed to provide comparable isolation to those present on mobile operating systems. 
The problem with this approach as I see it, is that is does not align well with the expectation of (many) Linux users that wish to isolate their applications through apparmor. Not only does Linux not provide nearly the same infrastructure and common methods of sharing file like the share feature on Android/IOS, the use cases of most Linux applications are a lot broader since general purpose systems like most Linux Desktops are quite different from mobile systems. When the mobile approach for security boundaries is applied to apparmor profiles intended for general purpose systems, this leads back to the dreaded "OMG apparmor again" situation that we want to avoid.

Instead of forcing rather alien security boundaries of mobile systems into the apparmor profiles in this repository that are meant for desktop systems, I would suggest a more reasonable approach for normal desktop applications. Instead of trying to bind file access for applications like image viewers to specific directories like "Pictures" and preventing any user from viewing image files in e.g. `~/Downloads `or `~/MyWorkProjects `by default, read access should be granted to any directory located in places that commonly contain any user files like in [this](https://github.com/roddhjav/apparmor.d/pull/292/files#diff-4f173a2f11e37f7021d0b3e23a0ce55bb3f2a1a798ef0d16503b12c25472fa34) example.
The difference for the practical security the apparmor profiles provide with this approach compared to the more limited mobile approach is nearly non existent for most applications, at least when no network access is required. The main objective should be to prevent any easy sandbox escape, like it would be possible if write access to hidden files are granted. For applications that do not have network access, the number of non hidden directories in $HOME they can read is completely irrelevant to the provided security of the profile. 

If an image viewer is compromised there are two possible consequences: 
1. The apparmor profile provides enough isolation to prevent sandbox escape and no other system vulnerability like privilege escalation through kernel vulnerabilities are present to be exploited. If this holds true the number of accessible normal user files is irrelevant since they cannot be extracted or manipulated.
2. The apparmor profile allows sandbox escape or another system vulnerabilities allows the circumvention of apparmor. In this case the number of accessible normal user files is still irrelevant since all of apparmors isolation can be circumvented.

In consequence, requiring a strict isolation of normal user files (as in pictures, documents, music ...) only makes using the apparmor profiles harder while it provides no security benefits as long as the profile is already isolated to the local system, meaning no known sandbox escape or path for extracting files like network access. (I would also consider the absence of X11 access a requirement)

For applications that require network access, the current approach makes a lot more sense and when a browser only gets access to `~/Downloads` this is a lot more reasonable compared to restricting an image viewer to `~/Pictures` despite the lack of network access.

Consequently I suggest to change the defined isolation objective for profiles to include a more granular decision tree in regard to how much access to files can be permitted.

**Comment by roddhjav** [comment]
I had a bit of thought about it. And it is still a hard no. However...

A complete reply should be me pushing a security model for this project (but that is still a wip). For now, I will keep it short (and incomplete).

The main idea behind MAC, is not to target some specific attacks and to protect the application against them. But to protect the application confined against unknown attack, ie: **anything** (within the limit of apparmor capabilities, obviously apparmor is neither a firewall nor a sandbox manager). In other words, the best (and only) way to ensure protection against an unknown range of attack is to use and abuse the **principle of least privilege**. It is exactly what this project is doing. Consequently, a profile should respect the same general rule regardless of its allowed access (internet, IPC, caps...).

Regarding usability, Linux applications do **not** have security as compatibility requirement. Therefore, yes, trying to enforce any sort of security will always generate some compatibility issue. However, it does not mean that we do not have standards. The first being the XDG style directory definition. It is the option I selected in this project: you are free to organize yourself as you want... as long as you define your directories in apparmor. Is it a trade of regarding usability vs security? Yes. However, I am convinced it is the only way for Linux system to provide state-of-the-art security.

This is for the theory. Regarding this specific example, we are talking about user data. To the end user, these are much more sensitive than anything else on the system. So being extra careful with them and not allowing `owner @{HOME}/[^.]**` makes sense. Otherwise, the risk of such a construction is an issue that Android also has: [xkcd 1319](https://xkcd.com/1319/): 
![](https://imgs.xkcd.com/comics/authorization.png)

> Instead of trying to bind file access for applications like image viewers to specific directories like "Pictures" and preventing any user from viewing image files in e.g. ~/Downloads or ~/MyWorkProjects by default, read access should be granted to any directory located in places that commonly contain any user files like in this example.

It works this way already for most profiles. For example, [loupe](https://github.com/roddhjav/apparmor.d/blob/b9cfd787c84400b7b91f6f99471c4773490cdfed/apparmor.d/groups/gnome/loupe#L16) has full access over all labelled directories. While [vlc](https://github.com/roddhjav/apparmor.d/blob/b9cfd787c84400b7b91f6f99471c4773490cdfed/apparmor.d/profiles-s-z/vlc#L94-L98) allow a smaller list of these directories. Is it perfect? Probably no, but it is both fully user [configurable](https://apparmor.pujol.io/configuration/#personal-directories) and much more restrictive than a systematic `@{HOME}/[^.]**`.

The only difference is that the current `user-read` abstraction ensures directories are well-defined under label (variable) such as we can easily update/control them now and in the future.

>As stated in the project objectives, the profiles in this repository are designed to provide comparable isolation to those present on mobile operating systems.

Nitpicking, but I have never said this. I take inspiration from the Android security implementation that features (among other things) full system confinement. It does not mean this project's aim is to implement all Android features... The reading of their security model can be a reminder that Google is first and foremost an ads company.

**However**

Despite this theory, I acknowledge that it might be a pain to have to handle extra configuration. Even more as it is still an early stage of the project. Therefore, I am also for a pragmatic implementation:
- Ideally, all user directories should be correctly labelled so that we can easily control them.
- But a leaser restriction is still far better than nothing.

**In conclusion**

- I have rewritten the user abstractions (see [2fc2394](https://github.com/roddhjav/apparmor.d/commit/2fc2394bad4f89dc0e367b8bc45c63196e4ef259)). There is `user-read` and  `user-write` (they have `@{HOME}/[^.]**`). While `user-read-strict`, `user-write-strict` only allow directories covered by variables.
- If you create a PR for a profile that use `user-read` or `user-write`, I will most likely said no.
- You are very welcome to propose new user style variables for user directory not yet covered.

**Extra note**

This may be revisited later, especially as a future version of apparmor will add a `prompt` keyword that will allow us to ask confirmation from the user before access is given.

**Reply by beroal** [reply]
I did not know about `prompt`. While it is close to what I want, it requires double work. First, I open a file `$F` in a program; then I check that the opened file is `$F`. Why not choose `$F` one time in a system dialog?

**Reply by roddhjav** [reply]
Not sure I understand your question here. But the profile is trusted while the user is not really trusted. Therefore, apparmor will only ask confirmation to the user when the profile tell apparmor it can do it for some rule (ie: a restricted set of rules under prompt`)

**Reply by beroal** [reply]
My remark is about optimizing the UI. Say, the profile for a graphics editor contains the rule `prompt @{HOME}/{,**} rw`. A user still needs to specify a file `$F` in `$HOME` to edit. So the user does this by opening `$F` in the graphics editor.

BTW, I can't find anything about this “prompt” feature on the web. Do you have any additional info?

**Reply by roddhjav** [reply]
It is an upcoming feature (the kernel side of apparmor need to support it). The integration into apparmor.d will depends on how exactly the feature work.

**Comment by valoq** [comment]
> It works this way already for most profiles. For example, [loupe](https://github.com/roddhjav/apparmor.d/blob/b9cfd787c84400b7b91f6f99471c4773490cdfed/apparmor.d/groups/gnome/loupe#L16) has full access over all labelled directories. While [vlc](https://github.com/roddhjav/apparmor.d/blob/b9cfd787c84400b7b91f6f99471c4773490cdfed/apparmor.d/profiles-s-z/vlc#L94-L98) allow a smaller list of these directories. Is it perfect? Probably no, but it is both fully user [configurable](https://apparmor.pujol.io/configuration/#personal-directories) and much more restrictive than a systematic `@{HOME}/[^.]**`.

My suggestion would address this inconsistency and instead apply a rule to which restriction can/should be set by default when conditions are met. 
To quote the xkcd: "to read my email, take my money, and impersonate me" is not possible in the proposed scenario, as access to arbitrary user files is only granted when no network/extraction path is available.
For some reason the example profiles you pointed out already implement this as it seems that vlc (which is restricted to specific XDG directories) has network access while the less restricted loupe does not have any permissions that would allow file extraction.

> **However**
> 
> Despite this theory, I acknowledge that it might be a pain to have to handle extra configuration. Even more as it is still an early stage of the project. Therefore, I am also for a pragmatic implementation:
> 
>     * Ideally, all user directories should be correctly labelled so that we can easily control them.
> 
>     * But a leaser restriction is still far better than nothing.
> 
> 
> **In conclusion**
> 
>     * I have rewritten the user abstractions (see [2fc2394](https://github.com/roddhjav/apparmor.d/commit/2fc2394bad4f89dc0e367b8bc45c63196e4ef259)). There is `user-read` and  `user-write` (they have `@{HOME}/[^.]**`). While `user-read-strict`, `user-write-strict` only allow directories covered by variables.

This was basically what I intended to provide with https://github.com/roddhjav/apparmor.d/pull/292 to avoid the need for users to redefine this optional abstraction.
> 
>     * If you create a PR for a profile that use `user-read` or `user-write`, I will most likely said no.

My idea would be to add a simple condition for this abstraction before it is permitted:
- No network access
- No X11 access
- No full device access (or at least no terminal access)

If those conditions are met, there should be little reason not to allow this lesser restriction for the above mentioned reasons.
Not to mention it would also allow more consistent profiles.

**Comment by beroal** [comment]
I don't agree that all programs without internet access should be allowed to access every user file. Then any program can delete all my work. Instead, I want that a program has access only to files I choose and only for the time period I choose.

Android since the 4th version has been moving to this security policy. However, AppArmor does not have such a feature. Probably, because it was designed for a different use case.

BTW, surprisingly large number of application programs require internet access. For example, `mpv` can play URLs. IMHO, in this case, internet access should be given dynamically as I described above for file access.

**Comment by valoq** [comment]
> I don't agree that all programs without internet access should be allowed to access every user file. Then any program can delete all my work. 

The condition includes only read access, so it should not matter at all since no damage can be done (as long as apparmor holds)

> Instead, I want that a program has access only to files I choose and only for the time period I choose.

The problem is what to choose as a default, because the files you want a program to access are almost always different from the next users files.

Perhaps a solution may be to define different access levels for user files and allow automatically setting the desired permissions at install time. For example:

Option A) 
Allow access to extended XDG directories (like it is currently the case for many profiles)

Option B)
Allow access to all user files in $HOME and $MOUNTPOINTS (excluding hidden files)

Option C) 
Set custom access permissions

This could be combined with a definition of application classes like Document-Applications, Media-Applications, Internet-Applications

**Comment by curiosityseeker** [comment]
I sympathize with this idea. I suggest the following approach:

1. Allow read access according to the `user-read` abstraction. 
2. Deny access to critical files/folders by adding the `deny-sensitive-home` and `private-files-strict` abstractions.
3. Limit write access to the supported file formats - similarly like Ubuntu is doing it in their libreoffice profile:

```
#Defines all common supported file formats
#Some obscure ones we're excluded (mostly input)

#Generic
#.txt
@{libreoffice_ext} = [tT][xX][tT]
#All the open document format
@{libreoffice_ext} += {,f,F}[oO][dDtT][tTsSpPbBgGfF]
#.xml and xsl
@{libreoffice_ext} += [xX][mMsS][lL]
#.pdf
@{libreoffice_ext} += [pP][dD][fF]
#Unified office format
@{libreoffice_ext} += [uU][oO][fFtTsSpP]
#(x)htm(l)
@{libreoffice_ext} += {,x,X}[hH][tT][mM]{,l,L}
#.epub
@{libreoffice_ext} += [eE][pP][uU][bB]
#.ps (printing to file)
@{libreoffice_ext} += [pP][sS]

#Images
@{libreoffice_ext} += [jJ][pP][gG]
@{libreoffice_ext} += [jJ][pP][eE][gG]
@{libreoffice_ext} += [pP][nN][gG]
@{libreoffice_ext} += [sS][vV][gG]
@{libreoffice_ext} += [sS][vV][gG][zZ]
@{libreoffice_ext} += [tT][iI][fF]
@{libreoffice_ext} += [tT][iI][fF][fF]

#Writer
@{libreoffice_ext} += [dD][oO][cCtT]{,x,X}
@{libreoffice_ext} += [rR][tT][fF]

#Calc
@{libreoffice_ext} += [xX][lL][sStT]{,x,X,m,M}
@{libreoffice_ext} += [xX][lL][wW]
#.dif dbf
@{libreoffice_ext} += [dD][iIbB][fF]
#.tsv .csv
@{libreoffice_ext} += [cCtT][sS][vV]
@{libreoffice_ext} += [sS][lL][kK]

#Impress/Draw
@{libreoffice_ext} += [pP][pP][tTsS]{,x,X}
@{libreoffice_ext} += [pP][oO][tT]{,m,M}
#Photoshop
@{libreoffice_ext} += [pP][sS][dD]

#Math
@{libreoffice_ext} += [mM][mM][lL]

....

owner @{HOME}/**.@{libreoffice_ext} rwk,
owner @{MOUNTS}/**.@{libreoffice_ext} rwk,
```
This can also be implemented in a similar way for e.g., okular, evince, gwenview etc.

**Reply by valoq** [reply]
> I sympathize with this idea. I suggest the following approach:
> 
>     1. Allow read access according to the `user-read` abstraction.

This should depend on the other permissions within the profile. As long as the profile does not allow network access to exfiltrate the data, the amount of normal user data the app may access has no real effect on security.

What I find even more disturbing is the illusion of control the user-read-strict abstraction implies:
Either the user has all their files in the XDG directories that are accessible with user-read-strict, or the apparmor profiles will get in their way e.g. for files like `@{HOME}/myproject`. So the only ways to deal with this is to either expand the permissions or move the myprojects directory inside one of the predefined XDG directories. If all files are in the XDG directories, this allows access to the very same files that were accessible with the user-read abstraction that allows access to all non hidden files in home.

The point Alex is trying to make with this XDG abstraction does seem valid to me in theory, especially in terms of the idea of moving towards the mobile platform permission model. But the fact remains that Linux desktops are still far away from the infrastructure that enables this mobile platform permission model to become an effective and usable environment. Without things like a "share" button to send a file from one application to another, applications need common access to user files they may potentially have to access and
forcing files into predefined directories will not add anything in terms of security. 

> 
>     2. Deny access to critical files/folders by adding the `deny-sensitive-home` and `private-files-strict` abstractions.

The user-read abstractions already do not allow access to those files. They allow access only to files in @HOME that do not start with a dot.
https://github.com/roddhjav/apparmor.d/blob/c3de88e476c219f5b4aee064f6ec39399917b243/apparmor.d/abstractions/user-read#L11

Additionally, deny rules should also be avoided at all costs.
See the chapter on Enumerating Badness here: https://www.ranum.com/security/computer_security/editorials/dumb/

> 
>     3. Limit write access to the supported file formats - similarly like Ubuntu is doing it in their libreoffice profile:

This idea sounds quite promising, assuming it works reliably in apparmor. 
@roddhjav Did you consider this approach before? I am curious if there are any hidden issues with filtering file extensions like this.

**Reply by curiosityseeker** [reply]
> The user-read abstractions already do not allow access to those files. They allow access only to files in @HOME that do not start with a dot.

Yes, that's true. My mistake.

> > 3. Limit write access to the supported file formats - similarly like Ubuntu is doing it in their libreoffice profile:
> > ```
> 
> This idea sounds quite promising, assuming it works reliably in apparmor. @roddhjav Did you consider this approach before? I am curious if there are any hidden issues with filtering file extensions like this.

Well, it has obviously worked in Ubuntu for several years. And I've been using this approach as well in some profiles for quite a while. The only problem is that you'll get a lot of noise by denials in `aa-log` for the non-@{libreoffice_ext} files. I don't know how to avoid this.

EDIT: Your comment about `deny` rules is certainly correct. Nevertheless, they can limit the risks of (potentially) too loose permisions - and remember, we're talking about read permissions here. Another useful side effect is that `deny` rules silence logging in `aa-logprof` or `aa-log`.  Unfortunately, something like `deny @{HOME}/[^**.@\{libreoffice_ext\}] rw,` does not work, and it also blocks access to, e.g., `@{user_config_dirs}` and `@{user_cache_dirs}`. 

Hint: If you want to test this you should not add above code for the @{libreoffice_ext} definitions in the .../local/libreoffice file. The apparmor_parser won't accept that for whatever reason. You have to add it to the profile itself.

**Comment by valoq** [comment]
https://apparmor.pujol.io/concepts/

> What to confine and why?
> 
> We take inspiration from the [Android/ChromeOS Security Model](https://arxiv.org/pdf/1904.05572v2.pdf), and we apply it to the Linux world. Modern [Linux security distributions](https://clip-os.org/en/) usually consider an immutable core base image with a carefully selected set of applications. Everything else should be sandboxed. Therefore, this project tries to confine all the core applications you will usually find in a Linux system: all systemd services, xwayland, network, bluetooth, your desktop environment, etc. **Non-core user applications are out of scope as they should be sandboxed using a dedicated tool (minijail, bubblewrap, toolbox, etc).**

Id like to bring this up again because reading it again this differentiation seems quite arbitrary and confusing. 
The differentiation between confinement and sandbox does not seem to be supported by any literature I found on the topic and there are many references to apparmor that explicitly mention its use for application "sandboxing". The term is overloaded however so I will avoid it for now. What we are talking about can be generally described as process isolation and frankly I don't see why there would be a differentiation between services, core tools and user applications (browser, media players etc.) or at least not in regards of terminology.

This project also includes many "Non-core user applications" already, which makes me wonder if the above statement is simply outdated?
Additionally, I very much disagree with the mentioned alternatives. I spend about a decade researching this topic by now, started supporting the development of firejail in 2016, before I moved on to bubblewrap and container isolation. Based on my experience during that time Apparmor is still the best solution by far to isolate user applications. With the additional issues regarding firejails use of suid and bubblewraps (and other container tools) use of user namespaces that breaks modern browsers internal sandboxing, Id even say its the only sane alternative that will at least not make things worse. I think the profiles we have already shows quite well that apparmor can fullfill these use cases as well and it does not require any additional tooling, especially not container technology (namespaces). If someone has an example where namespaces can provide anything important that apparmor can not, Id be interested to hear about it.

What I really want to mention though is that this projects supports a number of use cases that differ from the descriptions.

Personally I use this project very differently from what you have envisioned according to the documentation. Most profiles that address system wide services are set in complain mode on my system, which makes sure everything they aren't expected to access, gets logged by auditd and sent to a read-only log server. The only profiles I enforce, are those for applications that actually face relevant attack surface, e.g. applications that open complex and untrusted data. The reason behind this is mainly the administrative efforts for maintaining profiles, which makes sense to me in case of the high risk applications and that is probably true for most users Id wager. Point is, the description of this project seems abit narrow when it already supports way more use cases.

**Reply by beroal** [reply]
> The differentiation between confinement and sandbox does not seem to be supported by any literature I found on the topic and there are many references to apparmor that explicitly mention its use for application "sandboxing".

I believe that sandboxing allows to substitute one resource for another from the point of view of the sandboxed program. For example, each command that the program attempts to execute against `/tmp/x` is executed against `/tmp/app/x`. AppArmor can't do this, but the kernel namespace feature can. Without this feature, in order to isolate a program, you need to collect the set of filenames that the program may access in `/tmp`. This is a nightmare because filenames are generated randomly using various algorithms. If the filename  patterns of two programs intersect, you just can't isolate these programs from each other. There are similar problems with `/dev/shm` and `/dev/tty*`.

> If someone has an example where namespaces can provide anything important that apparmor can not, Id be interested to hear about it.

- AppArmor can't substitute one resource for another (see above).
- A user can't give permissions.
  - A user can't give permissions to an editor to access a file that the user wants to edit.
  - On real multiuser systems, if a user wants to use a new program, the root gives the program permissions, and then the program has these permissions when executed under any user. This is dangerous for other users as the program can be started implicitly, for example, via `xdg-open`.

**Reply by valoq** [reply]
> I believe that sandboxing allows to substitute one resource for another from the point of view of the sandboxed program. For example, each command that the program attempts to execute against /tmp/x is executed against /tmp/app/x. AppArmor can't do this, but the kernel namespace feature can


Im not sure I follow what you mean here. Nor do I see how namespaces help with this.

As for filenames, with #906 most real problems are addresses by allowing file types instead. The Music player gets access to all media files but no office docs. That provides a easy to maintain solution.

The biggest issue I still have with namespaces is that they provide abstractions for resources provided by the kernel, but do not isolate the kernel itself. I need to look into apparmor again, but it is likely able to isolate the kernel itself way more, which is the main attack vector after all.

**Comment by roddhjav** [comment]
They have been a lot of work since I wrote this quick description. A more in depth model is coming, and a draft of it should be on-line soon.

But quickly, yes, the use case is various. For example:
- From: The profiles come pre installed with the distribution, users should not even know apparmor is enabled.
- To: FSP, RBAC, even your interactive terminal is confined, you need special role for administrative tasks: https://github.com/roddhjav/play

> If someone has an example where namespaces can provide anything important that apparmor can not, Id be interested to hear about it.

**Apparmor has some limitation/drawback**
- There are no (or not enough) LSM hooks for thing like cgroup, pbf, or io_uring, therefore apparmor (and SELinux) cannot control them as they should.
- You cannot filter by syscall (like seccomp does),
- You don't hide your process from other (like a cgroup)
- It is not a firewall
- In usage sandboxes are different:
  - Image based container/VM: you fully hide yourself from others, (note: it can actually be an issue sometime)
  - It applies very well to program that do not communicate much with other (eg: http only for server,  xdg-desktop-portal...).

**However, this does not matter that much.**

Apparmor is only a tool, and as a tool it can be used for various things. Eg: *(copy pasting from draft arch)*

1. **In application sandboxing:** separation of privilege, least privilege principle, within different process of the application itself. It can be implemented using tools such as: Landlock, bwrap, and Apparmor.
The purpose is to separate highly privileged code from the rest of the application: Eg: pacman, VLC, Chrome sandboxing, **[loupe](https://github.com/roddhjav/apparmor.d/blob/e490a11c1a2ecfadd2cbc0759d77f4706bc2ee61/apparmor.d/groups/gnome/loupe#L62C2-L81)**.

1. **Application sandboxing:** Isolate the application from the rest of the system with tools such as. In this context, "sandboxing" does not refer to a special technology, but to the general concept. It can be implemented in various ways: trought VM (Qemu/KVM, Firecracker, Cloudhypervisor, Kata Containers), or container (Docker, gVisor, Flatpak, Snap) with different level of isolation and integration with the rest of the system.

1. **Confined user:** The user is confined in its own environment. Limiting what they can do -> https://github.com/roddhjav/play

1. **System confinement:** separation of privilege, least privilege principle, within all part of the core system. For example, Gnome is constituted by about 50 small services running in the background. Each of them are confined independently with granular access to other part of the DE.

<img width="685" height="571" alt="arch2" src="https://github.com/user-attachments/assets/1c03c6dc-faf0-47cb-8f87-47cd11b28d56" />

More regarding the general architecture: https://lssna2025.sched.com/event/1zalf/lessons-learned-while-making-an-apparmor-play-machine-alexandre-pujol-linagora

**Reply by valoq** [reply]
> In application sandboxing: separation of privilege, least privilege principle, within different process of the application itself. It can be implemented using tools such as: Landlock, bwrap, and Apparmor.

Not sure why you would put apparmor and bwrap in this category. I see that you can use partial privilege separation when an application calls another and you isolate the called application in a different profile, but this isn't sandboxing within the same application. The term I often see and prefer for this would be native sandboxing using things like Landlock and Seccomp, optionally in combination with a broker architecture for separate processes. [Sandboxed-API](https://github.com/google/sandboxed-api) also fits in this category.


> The purpose is to separate highly privileged code from the rest of the application: Eg: pacman, VLC, Chrome sandboxing, [loupe](https://github.com/roddhjav/apparmor.d/blob/e490a11c1a2ecfadd2cbc0759d77f4706bc2ee61/apparmor.d/groups/gnome/loupe#L62C2-L81).

I'm not aware of any internal sandboxing in VLC. However a good examples for this category would be [OpenSSH](https://www.usenix.org/legacy/event/sec03/tech/full_papers/provos_et_al/provos_et_al_html/) which supported this kind of privilege separation more then two decades ago already.


The graphic is something I have stared at for some time now (I also saw your recent talk about the Apparmor Play machine).
The separation between DGM services and (systemd?) services seems strange too. Why would we separate this when describing these services? You can sandbox all of these, for example using the integrated tools for systemd services and then isolate them again using Apparmor. I don't see why its two categories.

And lastly all of the "Sandboxed App" in the graphic is what I could call containerized applications, since that is what they use. They all build runtime environments and execute applications inside that were not designed to be run inside containers (which may not be an issue in itself). Point is, they all follow a very particular security model that separates the application and all the data it can legitimately access from the rest of the system. Real sandboxed applications aka natively sandboxed programs have a much better security model in comparison, since they can, within their own application separate complex and dangerous functions into separate and locked down child processes that never get to see all data. Also see this [talk](https://www.youtube.com/watch?v=2e91cEzq3Us).

**Reply by valoq** [reply]
The different categories are probably best described and separated by their security model:

**In application isolation/native sandbox:**
Implements a separation of privileges within the application source code itself.
This can be done using more restricted child processes or self restriction before executing dangerous functions (like parsing files).
Perhaps this can be further separated into two approaches: The simplified (single process) self restriction for one time execution applications that open a file descriptor, enforce self sandboxing and then parse the data (execute dangerous code). Examples could be [zathura](https://github.com/pwmt/zathura/blob/cbfa776ad1db82ba2d440e36eeef6c6ad0d4cc00/zathura/main-sandbox.c#L50), [file](https://github.com/file/file/blob/74e451be9d9e86af5e06e9795403017f8c6833ba/src/file.c#L368) and [ouch](https://github.com/ouch-org/ouch/pull/723/files)
The second variant being broker architectures like implemented in OpenSSH, Chromium and Firefox, where dangerous functions like pre-authentication and renderers are split off into child processes while the main process keeps normal DAC privileges.

The primary attribute of the thread model here is that the application may potentially have access to all files but the restricted/dangerous code only receives access to user chosen files (e.g. uploading a file after selecting it in the file chooser) through the broker API. In the context of browsers, this also isolates different web pages from each other. If a renderer process is compromised, the confidentiality of other files the same application has access to is not impacted, which is the main difference to the second model:

**Containerized application sandbox:**
Implements a runtime environments with all dependencies and executes the (unmodified) application inside.
This isolates the application from all resources it does not need access to, but has to allow access to all files it could potentially access. Compromising this application will compromise all files within the container, which is often all user files or at least all files of a specific type (e.g. all media files).

**Service isolation sandbox:**
This is basically the same as the containerized application sandbox but tailored to services/daemons, including things like GDM, pipewire etc.
The difference to the second type is that it often uses the DAC model as well since most of these services run as their own user, thereby approaching the android security model. Additional difference is that services usually do not need access to user resources like files and can therefore be restricted more effectively then containerized applications
